src/d3a/d3a_core/simulation.py

```diff
@@ -169,7 +170,11 @@ def _init(self, slot_length_realtime, seed, paused, pause_after, redis_job_id):
             log.info("Random seed: {}".format(random_seed))
 
         self.area = self.setup_module.get_setup(self.simulation_config)
+        # TODO: discuss whether the following circular reference is a good idea (probably not)
+        #  it is needed for being able to update the whole grid_tree from the devices
```

Comment:
I am fairly certain that we'll unfortunately regret the circular reference in the future. In this case, even though it is an antipattern, I think that having a "singleton" is much better than having to deal with strong references. I want to see the calling code of this class though, if I find a less intrusive way than a singleton I'll edit this comment.
If you decide to keep this version for whatever reason, I would at least make the references weak, in order to not prohibit areas that are deleted to be garbage-collected https://docs.python.org/3/library/weakref.html.

Reply:
Interesting read. Thanks! But I fear that if not implemented correctly we still run into the memory leak hole (talking about the implementation of the `finalize` function) .
I would rather try to implement the Global_Statitics as a singleton. Should be straight forward and I already thought about that but remembered ("No, singletons are bad!").